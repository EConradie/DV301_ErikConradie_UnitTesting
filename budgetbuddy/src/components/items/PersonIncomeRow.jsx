import React from 'react'

function PersonIncomeRow(props) {

  const {person, index} = props;

  return (
    <div className='card card-row p-2 mb-2 row-income' data-testid="income-card">
        <span>
            <div  className='avatar' aria-label='icon'>
                {person.icon}
            </div>
            <p aria-label ='name'>{index}. {person.name}</p>
        </span>
        <p aria-label='income' className='amount'>R {person.salary.toFixed(2)}</p>
    </div>
  )
}

export default PersonIncomeRow