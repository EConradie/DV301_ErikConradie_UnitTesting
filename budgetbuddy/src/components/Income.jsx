import React, { useState } from "react";
import PersonIncomeRow from "./items/PersonIncomeRow";
import { Button, Form } from "react-bootstrap";
import { iconOptions } from "../utils";
import {
  calculateTaxBracket,
  calculateTaxAmount,
} from "../components/functions/TaxCalculations";

function Income({ income: initialIncome, handleNewIncome }) {
  const [income, setIncome] = useState([...initialIncome]);
  const [icon, setIcon] = useState("-");
  const [name, setName] = useState("");
  const [salary, setSalary] = useState("");

  const handleAddMember = (event) => {
    event.preventDefault();

    const salaryNumber = Number(salary);
    const bracket = Number(calculateTaxBracket(salaryNumber));
    const taxAmount = Number(calculateTaxAmount(salaryNumber, bracket));

    if (name && salaryNumber) {
      const newIncome = { icon, name, salary: salaryNumber, bracket, taxAmount };
      setIncome([...income, newIncome]);
      handleNewIncome(newIncome);

      setIcon("-");
      setName("");
      setSalary("");
    }
  };

  return (
    <div data-testid='income-component'>
      <h3>Household Monthly Salaries</h3>
      <Form onSubmit={handleAddMember}>
        <div className="form-row">
          <Form.Select aria-label="icon" name="icon" value={icon} onChange={(e) => setIcon(e.target.value)} autoComplete="off">
            <option disabled value="-">-</option>
            {iconOptions.map((icon, index) => (
              <option key={index} value={icon}>{icon}</option>
            ))}
          </Form.Select>
          <Form.Control
            type="text"
            aria-label="name"
            name="name"
            placeholder="Member Name"
            value={name}
            onChange={(e) => setName(e.target.value)}
            autoComplete="off"
          />
          <Form.Control
            type="number"
            aria-label="income"
            name="income"
            step="0.01"
            placeholder="0.00"
            value={salary}
            onChange={(e) => setSalary(e.target.value)}
            autoComplete="off"
          />
          <Button type="submit" className="add-income">
            Add
          </Button>
        </div>
      </Form>
      <div className="income-outer hide-scroll">
        {income.map((item, index) => (
          <PersonIncomeRow key={index} index={index + 1} person={item} />
        ))}
      </div>
    </div>
  );
}

export default Income;
