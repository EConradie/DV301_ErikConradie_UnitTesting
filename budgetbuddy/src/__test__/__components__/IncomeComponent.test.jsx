import { render, screen } from "@testing-library/react"
import Income from "../../components/Income"
import PersonIncomeRow from "../../components/items/PersonIncomeRow";
import userEvent from "@testing-library/user-event";

test("Test default state of Income Component", () => {
  render(<Income income={[]} />);

  var fullComponent = screen.getByTestId("income-component");
  var title = screen.getByText("Household Monthly Salaries");
  var icon = screen.getByLabelText("icon");
  var name = screen.getByLabelText("name");
  var income = screen.getByLabelText("income");
  var button = screen.getByRole("button");

  expect(title).toBeInTheDocument();
  expect(icon).toBeInTheDocument();
  expect(name).toBeInTheDocument(); 
  expect(income).toBeInTheDocument();
  expect(button).toBeInTheDocument();

  expect(fullComponent).toMatchSnapshot();
});

test("Test if income cards display when 2 people are added", () => {
  var personMock = [
    {
      name: "Income 1",
      salary: 1000,
      icon: "ðŸ‘©",
      bracket: 0.0,
      taxAmount: 0.0,
    },
    {
      name: "Income 2",
      salary: 1000,
      icon: "ðŸ‘©",
      bracket: 0.0,
      taxAmount: 0.0,
    }
  ];

  render(
    <div>
      {personMock.map((person, index) => (
        <PersonIncomeRow key={index} person={person} />
      ))}
    </div>
  );

  var iconDisplay = screen.getAllByLabelText("icon");
  var nameDisplay = screen.getAllByLabelText("name");
  var incomeDisplay = screen.getAllByLabelText("income");

  var incomeCards = screen.getAllByTestId("income-card"); 

  incomeCards.forEach((element, index) => {

    expect(element).toBeInTheDocument();
    expect(element).toHaveTextContent(`R ${personMock[index].salary.toFixed(2)}`);
    expect(element).toHaveTextContent(personMock[index].name);
    expect(element).toHaveTextContent(personMock[index].icon);

  })

  expect(iconDisplay).toHaveLength(2);
  expect(nameDisplay).toHaveLength(2);
  expect(incomeDisplay).toHaveLength(2);    



});

test("Test adding of a new income", async () => {
  var user = userEvent.setup();

  var incomeMock = [];

  const handleNewIncome = (newIncome) => {
    incomeMock = [...incomeMock, newIncome];
  };

  render(<Income income={incomeMock} handleNewIncome={handleNewIncome}  />);

  var iconInput = screen.getByLabelText("icon");

  await user.selectOptions(iconInput, "ðŸ‘©")
  expect (iconInput).toHaveValue("ðŸ‘©");

  var nameInput = screen.getByLabelText("name");

  await user.click(nameInput)
  await user.keyboard("test")

  expect(nameInput).toHaveValue("test");

  var incomeInput = screen.getByLabelText("income");

  await user.click(incomeInput)
  await user.keyboard("1000")

  expect(incomeInput).toHaveValue(1000);

  var button = screen.getByRole("button");

  await user.click(button)

  expect(incomeMock.length).toBe(1)
  expect(incomeMock[0].icon).toBe("ðŸ‘©")
  expect(incomeMock[0].name).toBe("test")
  expect(incomeMock[0].salary).toBe(1000)
})
