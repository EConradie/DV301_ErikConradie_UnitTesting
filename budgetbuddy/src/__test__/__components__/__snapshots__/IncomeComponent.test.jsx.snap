// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default state of Income Component 1`] = `
<div
  data-testid="income-component"
>
  <h3>
    Household Monthly Salaries
  </h3>
  <form
    class=""
  >
    <div
      class="form-row"
    >
      <select
        aria-label="icon"
        autocomplete="off"
        class="form-select"
        name="icon"
      >
        <option
          disabled=""
          value="-"
        >
          -
        </option>
        <option
          value="👩"
        >
          👩
        </option>
        <option
          value="👱‍♂️"
        >
          👱‍♂️
        </option>
        <option
          value="🐶"
        >
          🐶
        </option>
        <option
          value="🐱"
        >
          🐱
        </option>
        <option
          value="🐰"
        >
          🐰
        </option>
        <option
          value="🐻"
        >
          🐻
        </option>
        <option
          value="🦁"
        >
          🦁
        </option>
        <option
          value="🐷"
        >
          🐷
        </option>
        <option
          value="🐮"
        >
          🐮
        </option>
      </select>
      <input
        aria-label="name"
        autocomplete="off"
        class="form-control"
        name="name"
        placeholder="Member Name"
        type="text"
        value=""
      />
      <input
        aria-label="income"
        autocomplete="off"
        class="form-control"
        name="income"
        placeholder="0.00"
        step="0.01"
        type="number"
        value=""
      />
      <button
        class="add-income btn btn-primary"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
  <div
    class="income-outer hide-scroll"
  />
</div>
`;
